src/files.ts

```diff
@@ -187,19 +187,48 @@ export async function collectIgnoreRules(
     // Check if symlink and exclude if requested
     if (!fileStats || (fileStats.isSymbolicLink() && !symlinksEnabled) || fileStats.isFile()) return [];
 
-    // Find ignore files inside this directory
+    // Parse top-level .snyk file inside this directory
+    const topLevelSnykIgnores = parseFileIgnores(`${folder}/${DOTSNYK_FILENAME}`);
+
+    // Find ignore files inside this directory, excluding
+    // paths that are ignored by top-level .snyk exclude rules.
     const localIgnoreFiles = await fg(
       IGNORE_FILES_NAMES.map(i => `*${i}`),
       {
         ...fgOptions,
         cwd: folder,
         followSymbolicLinks: symlinksEnabled,
+        ignore: topLevelSnykIgnores,
       },
     );
-    // Read ignore files and merge new patterns
-    return union(...localIgnoreFiles.map(parseFileIgnores));
+
+    // Read ignore files and merge patterns.
+    const localIgnoreRules = union(...localIgnoreFiles.map(parseFileIgnores));
+
+    // Final array of rules after processing to return.
+    const rulesAfterProcessing: string[] = [];
+
+    for (const rule of localIgnoreRules) {
+      // Identify and remove negative rules that are superseded by top-level .snyk rules.
```

Comment:
q: I wonder if we could make this simpler and more resilient by inverting the logic, similar to how scm-bundle-store does it. I'm not completely familiar with how the logic here works though. I'm thinking:
- Collect the .gitignore rules
- Apply the .gitignore rules (positive and negative)
- Apply the .snyk excludes

This could be implemented by returning an object with the different kinds of rules from this function, and applying them in sequence in the `collectBundleFiles` function.

The existing design where we force everything into a single set of ignore rules seems contrary to the domain logic of having .snyk excludes take precedence, and we are hacking around it with this rule-filtering logic.

Reply:
I was on the fence about this for a while, but eventually settled on this approach because 1) it is simple - we keep the same behaviour as before, where the "collect ignore rules" function returns a single array of rules to be used, and 2) returning different sets of rules and applying them in sequence requires either yet another pass of `searchFiles`, or significant rewriting of the file collection logic, neither of which seems worth it to me.